tests/mr/device/mr_multithreaded_tests.cpp

```diff
@@ -195,34 +196,41 @@ void allocate_loop(rmm::mr::device_memory_resource* mr,
       RMM_CUDA_TRY(cudaEventRecord(event, stream.value()));
       allocations.emplace_back(ptr, size);
     }
+    allocations_ready.notify_one();
   }
+  // Work around for threads going away before cudaEvent has finished async processing
+  cudaEventSynchronize(event);
 }
 
 void deallocate_loop(rmm::mr::device_memory_resource* mr,
                      std::size_t num_allocations,
                      std::list<allocation>& allocations,
                      std::mutex& mtx,
+                     std::condition_variable& allocations_ready,
                      cudaEvent_t& event,
                      rmm::cuda_stream_view stream)
 {
-  for (std::size_t i = 0; i < num_allocations;) {
-    std::lock_guard<std::mutex> lock(mtx);
-    if (allocations.empty()) { continue; }
-    i++;
+  for (std::size_t i = 0; i < num_allocations; i++) {
+    std::unique_lock lock(mtx);
+    allocations_ready.wait(lock, [&allocations] { return !allocations.empty(); });
```

Comment:
FWIW, I suspect that the deadlock we've observed comes from the cudaevent stuff (and that this is squashing a much less likely to be observed deadlock [still possible I guess]).

Reply:
I agree, the cudaevent issue was the primary cause but since we identified another locking issue it felt appropriate to correct both.